### ISSUE TITLE
Possible race condition in RemoteRenderer

### ISSUE THREAD

**Issue by avl** [dotnet/aspnetcore#43621]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We have a Blazor Server application where we have intermittent crashes in the blazor javascript.

We're suspecting a possible race condition bug somewhere in the server side Blazor-rendering code.

The error message in the browser console is this:

```
Uncaught (in promise) Error: No element is currently associated with component 1174470854
    at ie.updateComponent (blazor.server.js:1:20119)
    at blazor.server.js:1:114736
    at xn.processBatch (blazor.server.js:1:115108)
    at kt.<anonymous> (blazor.server.js:1:129290)
    at blazor.server.js:1:72016
    at Array.forEach (<anonymous>)
    at kt._invokeClientMethod (blazor.server.js:1:72002)
    at kt._processIncomingData (blazor.server.js:1:70044)
    at connection.onreceive (blazor.server.js:1:64447)
    at o.onmessage (blazor.server.js:1:48781)
```

Whenever this happens, we observe that the buffer in the method WriteBatchBytesAsync, in RemoteRenderer.cs, is corrupted while SendAsync is running. In fact, we can observe by doing debug-printouts, that the invalid id 1174470854 is actually the first 4 bytes of the corrupted data.

We tried modifying the aspnetcore code in the method WriteBatchBytesAsync in RemoteRenderer.cs like this:

```csharp
                var correct = new ArraySegment<byte>(pending.Data.Buffer.ToArray(),0,pending.Data.Count);
                var segment = new ArraySegment<byte>(pending.Data.Buffer, 0, pending.Data.Count);

                await _client.SendAsync("JS.RenderBatch", pending.BatchId, segment);

                if (!correct.SequenceEqual(segment))
                    Console.WriteLine("Segment corrupted!");

```
(We've added the 'correct' variable)

Doing this prints "Segment corrupted!" when the error occurs.

By using the 'correct' variable in the SendAsync call, causes the problem to go away.

We would appreciate any tips or hints on how to move forward. Doing the ToArray() is not free, so doing that is probably not a good fix.

Also, there is of course the possibility that the corruption occurs because of something our app is doing. 

We have noticed that the buffer that is received by the browser javascript has some corruption (5 first bytes are wrong), but the buffer after "await SendAsync" returns is almost completely overwritten. 

The behaviour seems to be as if there is a race that causes the pending UnacknowledgedRenderBatch to be overwritten/reused.

Would appreciate any hints/pointers on things to try. We are very motivated to get to the bottom with this :-) .

### Expected Behavior

Regarding expected behaviour, naturally we hoped to not have error messages from the javascript :-) .

### Steps To Reproduce

Unfortunately I can't share the customer-project where the problem occurs.

At this time I do not expect microsoft to be able to solve this issue for us.

But I hope for some pointers that might make it possible for us to identify the problem in detail and fix it ourselves! (While of course sharing any fix we find!)

### Exceptions (if any)

We get this exception from the C# aspnet server code:

```
warn: Microsoft.AspNetCore.Components.Server.Circuits.RemoteRenderer[100]
      Unhandled exception rendering component: Error: No element is currently associated with component 1174470854
      System.InvalidOperationException: Error: No element is currently associated with component 1174470854
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.InvokeRenderCompletedCallsAfterUpdateDisplayTask(Task updateDisplayTask, Int32[] updatedComponents)
fail: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
      Unhandled exception in circuit 'yQu2H9UAQ886H5MlQYhUqmkTL44emgAzXvDCABclojU'.
      System.AggregateException: One or more errors occurred. (Error: No element is currently associated with component 1174470854)
```

But it doesn't help much, since the problem actually occurs much earlier, as explained above.

### .NET Version

6.0.2

### Anything else?

We have tried various .NET versions but problem seems to exist with 6.0.2 and later. We haven't tried previous versions.

**Comment by javiercn** [comment]
Potential context:
* https://github.com/dotnet/aspnetcore/issues/5592
* https://github.com/dotnet/blazor/pull/1666

**Comment by javiercn** [comment]
@avl thanks for contacting us.

* Are you able to reproduce the issue "consistently"?
* Could you collect the debug logs to see if that helps shed some clarity on what might be going on?
* Does this happen when the app launches, after a few minutes, or at any point?
* Could you capture the call stack at the time you think the segment is corrupted?

Looking at the relevant code I do not see anything immediately obvious, so we need more data to hopefully try and identify what is going on.

**Comment by avl** [comment]
Hello! 

Thanks for your response!

Working in Europe, so ending my work day very soon now. Will provide more information tomorrow!

I can reproduce the issue sort of consistently. It requires clicking in our app for approximately 2 minutes. It's intermittent and random, but there are some views which seem to provoke it more oftenly. It only happens on some computers. One of them is a 16 core Ryzen 5950X, the other is a several years old laptop. We don't understand what is special with these two computers.

Which debug logs are you interested in? I've tried enabling "trace" logging level, but not much is logged for these parts of the code. I've therefore added a lot of new logging myself.

I'm not sure if the issue has ever happened on the first render. Possibly not, but my feeling is that the issue doesn't depend on how long the server has been running.

I will look into capturing a call stack.

Right now I tried switching the ArrayBuilder created in UpdateDisplayAsync into a local (per thread) pool created using ArrayPool<byte>.Create();  This seems to have solved the problem also. 

I almost looks as if something, somewhere, is keeping a buffer created from an ArrayPool, for a little too long. It doesn't seem to be the code in RemoteRenderer though.

This is actually quite fun to debug, sort of a little detective story :-) .

**Comment by javiercn** [comment]
@avl the ASP.NET Core logs. At least `Microsoft.AspNetCore.Components` and `Microsoft.AspNetCore.SignalR`. You can set the options in appSettings.json (beware that appsettings.development.json might override the log levels).

If you did things correctly, you should see a ton more log messages (with dbug) in them, for example, every time an event or a render happens on the server.

> I almost looks as if something, somewhere, is keeping a buffer created from an ArrayPool, for a little too long. It doesn't seem to be the code in RemoteRenderer though.

This is my theory too, I am trying to figure out if we can leverage the logs to see the sequence of events that cause it.

Also capture the logging on the browser:
For SignalR
https://docs.microsoft.com/en-us/aspnet/core/blazor/fundamentals/signalr?view=aspnetcore-6.0#configure-signalr-client-logging-blazor-server

For Blazor Server specifically
```
<script src="_framework/blazor.server.js" autostart="false"></script>
<script>
Blazor.start({
  logLevel: 1
})
</script>
```

**Comment by avl** [comment]
Thanks for your feedback.

I've enabled logging, and reproduced the issue. I've had to redact the log files slightly, to be allowed to release them. I don't think this takes away anything from the ability to debug them.

I've tried to determine if we use ArrayPool in our own app. From what I can tell it is only used from within protobuf-net, which we use. A bug there might cause this issue. However, we do think the error has been triggered in our app before protobuf-net is used the first time, although we're not absolutely certain.

<details>
  <summary>Blazor Server js log</summary>
<pre>
  [2022-08-30T07:18:05.391Z] Information: Normalizing '_blazor' to 'https://localhost:4443/_blazor'.
Utils.js:159 [2022-08-30T07:18:05.392Z] Debug: Starting HubConnection.
Utils.js:159 [2022-08-30T07:18:05.392Z] Debug: Starting connection with transfer format 'Binary'.
Utils.js:159 [2022-08-30T07:18:05.392Z] Debug: Sending negotiation request: https://localhost:4443/_blazor/negotiate?negotiateVersion=1.
Utils.js:159 [2022-08-30T07:18:05.394Z] Debug: Selecting transport 'WebSockets'.
Utils.js:155 [2022-08-30T07:18:05.703Z] Information: WebSocket connected to wss://localhost:4443/_blazor?id=hLDPhy4XIVIObYYwp_bwLg.
Utils.js:159 [2022-08-30T07:18:05.703Z] Debug: The HttpConnection connected successfully.
Utils.js:159 [2022-08-30T07:18:05.703Z] Debug: Sending handshake request.
Utils.js:155 [2022-08-30T07:18:05.703Z] Information: Using HubProtocol 'blazorpack'.
Utils.js:159 [2022-08-30T07:18:05.704Z] Debug: Server handshake complete.
Utils.js:159 [2022-08-30T07:18:05.704Z] Debug: HubConnection connected successfully.
BrowserRenderer.ts:43 bn
Loggers.ts:27 [2022-08-30T07:21:54.084Z] Error: There was an error applying batch 411.
log @ Loggers.ts:27
BrowserRenderer.ts:44 Uncaught (in promise) Error: No element is currently associated with component -50134842
    at ie.updateComponent (BrowserRenderer.ts:44:13)
    at Renderer.ts:69:21
    at xn.processBatch (RenderQueue.ts:54:7)
    at kt.<anonymous> (Boot.Server.ts:123:17)
    at HubConnection.js:526:42
    at Array.forEach (<anonymous>)
    at kt._invokeClientMethod (HubConnection.js:526:25)
    at kt._processIncomingData (HubConnection.js:410:30)
    at connection.onreceive (HubConnection.js:41:52)
    at o.onmessage (WebSocketTransport.js:75:30)
Loggers.ts:27 [2022-08-30T07:21:54.095Z] Error: System.AggregateException: One or more errors occurred. (Error: No element is currently associated with component -50134842)
 ---> System.InvalidOperationException: Error: No element is currently associated with component -50134842
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.InvokeRenderCompletedCallsAfterUpdateDisplayTask(Task updateDisplayTask, Int32[] updatedComponents)
   --- End of inner exception stack trace ---
log @ Loggers.ts:27
Utils.js:159 [2022-08-30T07:21:54.095Z] Debug: Stopping HubConnection.
Utils.js:159 [2022-08-30T07:21:54.096Z] Debug: HttpConnection.stopConnection(undefined) called while in state Disconnecting.
Utils.js:155 [2022-08-30T07:21:54.096Z] Information: Connection disconnected.
Utils.js:159 [2022-08-30T07:21:54.096Z] Debug: HubConnection.connectionClosed(undefined) called while in state Disconnecting.
HttpConnection.js:92 Uncaught (in promise) Error: Cannot send data if the connection is not in the 'Connected' State.
    at bt.send (HttpConnection.js:92:35)
    at kt._sendMessage (HubConnection.js:259:32)
    at kt._sendWithProtocol (HubConnection.js:266:21)
    at kt.send (HubConnection.js:279:34)
    at Object.beginInvokeDotNetFromJS (Boot.Server.ts:170:18)
    at w (Microsoft.JSInterop.js:212:37)
    at C.invokeMethodAsync (Microsoft.JSInterop.js:375:20)
    at WebRendererInteropMethods.ts:36:27
    at T (WebRendererInteropMethods.ts:51:108)
    at I (WebRendererInteropMethods.ts:34:10)
</pre>
</details>

<details>
  <summary>ASP Net Server log</summary>
<pre>
      Received hub invocation: InvocationMessage { InvocationId: "", Target: "BeginInvokeDotNetFromJS", Arguments: [ 177, , DispatchEventAsync, 1, [{"eventHandlerId":161928,"eventName":"click","eventFieldInfo":null},{"detail":1,"screenX":1806,"screenY":517,"clientX":701,"clientY":540,"offsetX":75,"offsetY":16,"pageX":701,"pageY":540,"button":0,"buttons":0,"ctrlKey":false,"shiftKey":false,"altKey":false,"metaKey":false,"type":"click"}] ], StreamIds: [  ] }.
dbug: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[202]
      Invoking instance method 'DispatchEventAsync' on instance '1' with callback id '177'.
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[5]
      Handling event 161928 of type 'MouseEventArgs'
dbug: Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime[3]
      Invocation of 'DispatchEventAsync' on reference '1' with callback id '177' completed successfully.
dbug: Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher[1]
      Received hub invocation: InvocationMessage { InvocationId: "", Target: "OnLocationChanged", Arguments: [ https://localhost:4443/supplyorder/MBAT.B26100-0002, True ], StreamIds: [  ] }.
dbug: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[208]
      Location changing to https://localhost:4443/supplyorder/MBAT.B26100-0002 in circuit 'UehuniAenM0FzAD8OBJFZHaD4JfAoHFmtke98CmAo6w'.
dbug: Microsoft.AspNetCore.Components.Server.Circuits.RemoteNavigationManager[2]
      Received notification that the URI has changed to https://localhost:4443/supplyorder/MBAT.B26100-0002 with isIntercepted=True
dbug: Microsoft.AspNetCore.Components.Routing.Router[2]
      Navigating to component XyzBlazorServer.Features.SupplyOrder.SupplyOrder in response to path 'supplyorder/MBAT.B26100-0002' with base URI 'https://localhost:4443/'
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 3 of type Microsoft.AspNetCore.Components.Routing.Router
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 4 of type Microsoft.AspNetCore.Components.Authorization.AuthorizeRouteView
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 5 of type Microsoft.AspNetCore.Components.Authorization.AuthorizeRouteView+AuthorizeRouteViewCore
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 6 of type Microsoft.AspNetCore.Components.LayoutView
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 7 of type XyzBlazorServer.Shared.MainLayout
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 8 of type XyzBlazorServer.Pages.RedactedInternalComponentLogin
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 9 of type XyzBlazorServer.Shared.NavMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 10 of type Microsoft.AspNetCore.Components.CascadingValue`1[RedactedInternalComponentLibrary.Components.DarkOrLightMode]
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12436 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12435 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12449 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.GenericTreePickerComponent`1[Xyz.RedactedDatabase.RedactedUnitDto]
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12434 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12393 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.RedactedInternalComponentGrid`1[Xyz.RedactedDatabase.RedactedSupplyOrderRowDto]
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12438 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12431 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12397 of type XyzBlazorServer.Components.XyzGroup
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12432 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12433 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12439 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12440 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12441 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12442 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12443 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12444 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12445 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12446 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12447 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12448 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12388 of type XyzBlazorServer.Components.XyzGroup
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12437 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12430 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12429 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.ContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 11 of type XyzBlazorServer.Components.XyzThemeToggler
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 12 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 13 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 14 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 15 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 16 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 17 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 18 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 19 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 20 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 21 of type Microsoft.AspNetCore.Components.Routing.NavLink
dbugMylog: Started constructing batch. batchId:411 Time: 232372089,10000002(us) thread: 55
: Microsoft.AspNetCore.Components.RenderTree.Renderer[3]
      Rendering component 22 of type RedactedInternalComponentLibrary.Components.ModalDialogHome
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[1]
      Initializing component 12462 (XyzBlazorServer.Features.SupplyOrder.SupplyOrder) as child of 22 (RedactedInternalComponentLibrary.Components.ModalDialogHome)
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12385 of type XyzBlazorServer.Features.SupplyOrder.SupplyOrderRows
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12386 of type XyzBlazorServer.Components.XyzHeaderFooterContent
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12387 of type XyzBlazorServer.Components.XyzRow
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12388 of type XyzBlazorServer.Components.XyzGroup
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12389 of type XyzBlazorServer.Components.XyzRow
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12390 of type XyzBlazorServer.Components.XyzCol
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12391 of type XyzBlazorServer.Components.XyzCol
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12392 of type XyzBlazorServer.Components.XyzRadioGroup`2[XyzBlazorServer.Features.SupplyOrder.SupplyOrderRows+Direction,XyzBlazorServer.Features.SupplyOrder.SupplyOrderRows+SupplyOrderRowsViewModel]
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12393 of type RedactedInternalComponentLibrary.Components.RedactedInternalComponentGrid.RedactedInternalComponentGrid`1[Xyz.RedactedDatabase.RedactedSupplyOrderRowDto]
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12394 of type XyzBlazorServer.Components.XyzCol
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12395 of type XyzBlazorServer.Components.XyzCol
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12396 of type XyzBlazorServer.Components.XyzUnitDropdownFormComponent
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12397 of type XyzBlazorServer.Components.XyzGroup
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12398 of type Microsoft.AspNetCore.Components.Forms.EditForm
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12399 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12400 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12401 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
Mylog: Enqueue buffer size: 261442 hashcode: 55129551 batchId: 411 Time: 232377744,5(us) thread: 55
Mylog: Before WriteBatchBytesAsync. batchId: 411 Time: 232377780,1(us) thread: 55
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12402 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
Mylog: Sending buffer size 261442 hashcode: 55129551 batchId: 411 Time: 232377866,5(us) thread: 55
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12403 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12404 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
Mylog: Rendercompleted Time: 232378136,7(us) thread: 55
Mylog: Disposing batch: 409 Time: 232378223,5(us) thread: 55
dbugMylog: ProcessPendingBatch, disposing entry.Data.Buffer with batchId: 409 hashcode: 44045295Time: 232378260,89999998(us) thread: 55
: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12405 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
Mylog: Rendercompleted Time: 232378367,1(us) thread: 55
Mylog: Sending buffer done. batchId: 411 Time: 232378424,6(us) thread: 41
Mylog: Segment corrupted! pending.Data.Buffer hashcode: 55129551 batchId: 411 Time: 232378489,1(us) thread: 41
dbugMylog: Disposing batch: 410 Time: 232378441,5(us) thread: 55
: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12406 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
Mylog: ProcessPendingBatch, disposing entry.Data.Buffer with batchId: 410 hashcode: 28534255Time: 232378617,3(us) thread: 55
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12407 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12408 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12409 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12410 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12411 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
dbug: Microsoft.AspNetCore.Components.RenderTree.Renderer[4]
      Disposing component 12412 of type XyzBlazorServer.Components.ContextMenus.SupplyOrderRowContextMenu
Mylog: Rendercompleted Time: 232379458(us) thread: 57
Mylog: Disposing batch: 411 Time: 232379496,5(us) thread: 57

</pre>
</details>
I had to shorten the log to fit here. The full log is available here: https://andersmusikka.se/log_to_microsoft.txt

**Comment by javiercn** [comment]
@avl if you are up for it, you could capture a memory dump when this happens (I believe it is possible with procdump to break on the exception and capture a dump).

That would allow you to see if something else is holding on to that buffer. It might happen that something is holding on to the buffer elsewhere, but returning it to the shared pool, then Blazor receives it and uses it, and the other component writes to it, introducing the error.

Another possibility was that we disposed it elsewhere (hence returning it to the pool) and then it got reused and we had a pending `SendAsync` but I have been looking at the code and I can't see that happening.

If I understand correctly, your repro does not involve the app being disconnected/reconnected, does it?

**Comment by javiercn** [comment]
@avl another way to investigate this will be to look at the trace log from the app after enabling the ArrayPoolEventSource (you can capture a trace with dotnet trace)

```
    [EventSource(Guid = "0866B2B8-5CEF-5DB9-2612-0C0FFD814A44", Name = "System.Buffers.ArrayPoolEventSource")]
```

I think that `--providers 0866B2B8-5CEF-5DB9-2612-0C0FFD814A44` or `--providers System.Buffers.ArrayPoolEventSource` will do the trick.

**Comment by javiercn** [comment]
https://docs.microsoft.com/en-us/dotnet/core/diagnostics/well-known-event-providers#systembuffersarraypooleventsource-provider

**Comment by avl** [comment]
Thanks for your suggestions! Sorry for the delay, was away on an internal meeting all day yesterday.

By tracing ArrayPoolEventSource I've found something!

Look at this: 


<HTML>
<TABLE><TR><TD>Event Name &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD> &nbsp;Time MSec</TD><TD>Process Name &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>Rest &nbsp;</TD></TR><TR><TD>System.Buffers.ArrayPoolEventSource/BufferRented &nbsp;</TD><TD>114,893.029</TD><TD>Process(30556) (30556)</TD><TD>HasStack=&quot;True&quot; ThreadID=&quot;33,020&quot; ProcessorNumber=&quot;0&quot; bufferId=&quot;34,071,888&quot; bufferSize=&quot;131,072&quot; poolId=&quot;35,342,034&quot; bucketId=&quot;13&quot; </TD></TR><TR><TD>System.Buffers.ArrayPoolEventSource/BufferRented &nbsp;</TD><TD>114,970.439</TD><TD>Process(30556) (30556)</TD><TD>HasStack=&quot;True&quot; ThreadID=&quot;33,020&quot; ProcessorNumber=&quot;0&quot; bufferId=&quot;34,071,888&quot; bufferSize=&quot;131,072&quot; poolId=&quot;35,342,034&quot; bucketId=&quot;13&quot; </TD></TR><TR><TD>System.Buffers.ArrayPoolEventSource/BufferReturned</TD><TD>114,970.745</TD><TD>Process(30556) (30556)</TD><TD>HasStack=&quot;True&quot; ThreadID=&quot;33,020&quot; ProcessorNumber=&quot;0&quot; bufferId=&quot;34,071,888&quot; bufferSize=&quot;131,072&quot; poolId=&quot;35,342,034&quot; </TD></TR><TR><TD>System.Buffers.ArrayPoolEventSource/BufferReturned</TD><TD>114,975.385</TD><TD>Process(30556) (30556)</TD><TD>HasStack=&quot;True&quot; ThreadID=&quot;24,960&quot; ProcessorNumber=&quot;0&quot; bufferId=&quot;34,071,888&quot; bufferSize=&quot;131,072&quot; poolId=&quot;35,342,034&quot; </TD></TR></TABLE>



It doesn't look right. Firstly, it looks like thread 33020 is renting the same buffer twice. Either

1) The ArrayPool is somehow returning the same buffer twice
2) The timestamps are slightly wrong, and thread 33020 just rents the buffer twice, returning it inbetween.

But the bigger problem is that thread 24960 is returning the buffer while thread 33020 has it rented.

The stack trace for the event at 114975.385 is the rendering code, the one where I detect the problem.

I have the stack trace for the event at 114970.745 (thread 24960), but I'm not sure how to get it out of visual studio. But basically it is like this:

ArrayPool.Shared.Return
MemoryBufferWriter.Reset
MemoryBufferWriter.Return
BlazorPackHubProtocolWorker.WriteMessage
BlazorPackHubProtocol.WriteMessage

There doesn't seem to be a way to copy-paste the stack trace out of the visualstudio nettrace-viewer.

**Comment by avl** [comment]
Ok, we have found the problem. 

Aspnetcore is not to blame! :-)

It turns out that one of our dependencies, ZstdSharp, had a bug where disposing a stream twice caused it to insert the same byte-array multiple times into ArrayPool<byte>.Shared . The reason why it only happened on two computers, is because they were the only ones actually using the feature which depends on ZstdSharp. But no one realized this could have anything to do with (what we thought were) aspnetcore-problems!

I have filed a report with ZstdSharp:
https://github.com/oleg-st/ZstdSharp/issues/8

Thank you so much for the debugging-help! Your hint to enable the ArrayPoolEventSource was the key.

Thanks again, and have a nice day!
